django/db/models/sql/compiler.py

```diff
@@ -1134,6 +1134,8 @@ def get_from_clause(self):
         """
         result = []
         params = []
+        # Copy `alias_map` to a tuple to prevent unintended alterations while
+        # compiling joins.
         for alias in tuple(self.query.alias_map):
```

Comment:
While I welcome the clarifying comment, and as someone that is not familiar with the intricacies of `alias_map`, could we go one step further and be explicit about what part of the dict this loop cares about?
```suggestion
        # Copy `alias_map` keys to a tuple to prevent unintended alterations
        # while compiling joins.
        for alias in tuple(self.query.alias_map.keys()):
```

Reply:
I think that it's pretty explicit already, just like any other usage of `dict.__iter__` we have in the code base where we don't explicitly call `.keys()` on it. I think the alteration is less about keys and more about the additions of more `JOINs` during the compilation of `JOINs`.

To give you some context @nessita `Query.alias_map` holds a reference to all the tables involved in the query. The first member is the base table (e.g. when you do `Author.objects.all()` it would be `BaseTable(Author._meta.db_table)`) and all the other members are `Join` instances resulting from `__` references in `filter`, `annotate`, and other `QuerySet` methods.

When we added support for `FilteredRelation` we decided that we wouldn't immediately add them to `alias_map` when `annotate` was called to avoid unnecessary `JOIN`s if the annotation was not referenced, instead it was decided that at `Join.as_sql` (the compilation phase) we would conditionally add it if necessary. Well this caused all sorts of issues mainly when the condition of the `FilteredRelation` was referring a table that wasn't already part of `alias_map`. Something like

```python
Author.objects.annotate(
    popular_books=FilteredRelation(
        "books", Q(books__rating__gte=F("books__publisher__popular_threshold")
    ),
    popular_books_cnt=Count("popular_books"),
).filter(popular_books_cnt__gte=2)
```

Would attempt to augment `alias_map` with a `JOIN` to `Publisher._meta.db_table` at compilation time when everything was already resolved. If we had the `tuple(self.query.alias_map)` from the get go it would have been clear from the early design of `FilteredRelation` that this wasn't going to fly.

Typing all that made me realize that there is a lot to unpack here and that a comment is certainly warranted 😅

Reply:
Thank you @charettes for the clear and detailed explanation. I think I understand it :-)

How would you recommend progressing this PR? I think the existing comment on its own is already an aid but perhaps some extra details to this same part of the logic would be appreciated by future readers.

Following "divide and conquer" approach I think I should merge this PR as is, and kindly invite you to, when possible and if your availability permits, propose a follow up PR with the extra commentary. What do you think?